lib/HLSL/DxilValidation.cpp

```diff
@@ -3180,11 +3184,19 @@ static void ValidateFunctionBody(Function *F, ValidationContext &ValCtx) {
       } break;
       case Instruction::AtomicCmpXchg:
       case Instruction::AtomicRMW: {
-        Type *T = cast<PointerType>(I.getOperand(AtomicRMWInst::getPointerOperandIndex())->getType())->getElementType();
+        Value *Ptr = I.getOperand(AtomicRMWInst::getPointerOperandIndex());
+        PointerType *ptrType = cast<PointerType>(Ptr->getType());
+        Type *elType = ptrType->getElementType();
         const ShaderModel *pSM = ValCtx.DxilMod.GetShaderModel();
-        if ((T->isIntegerTy(64)) && !pSM->IsSM66Plus())
+        if ((elType->isIntegerTy(64)) && !pSM->IsSM66Plus())
           ValCtx.EmitInstrFormatError(&I, ValidationRule::SmOpcodeInInvalidFunction,
                                       {"64-bit atomic operations", "Shader Model 6.6+"});
+        if (ptrType->getAddressSpace() != DXIL::kTGSMAddrSpace) {
+          ValCtx.EmitInstrError(&I, ValidationRule::InstrAtomicOpNonGroupshared);
+        } else if (GlobalVariable *GV = dyn_cast<GlobalVariable>(Ptr)) {
```

Comment:
Multiple thoughts:
1. Why `else`?  Shouldn't a tgsm constant global be a problem too?  Perhaps other validation should fire for that case, since it shouldn't be possible to declare constant groupshared in DXIL I think.
2. The pointer must ultimately come from a global TGSM.
3. The pointer may not itself be a `GlobalVariable`, but it could instead be a GEP of `GlobalVariable`, so you may need to drill through a GEP here.
4. If/when we support unions, I imagine we would also have to drill through pointer bitcast.

Reply:
1. The else case is what catches the tgsm constant global. It is possible to declare a constant groupshared and it validates fine.
2. The validation error below detects when a TGSMAddrSpace pointer doesn't point to a globalvariable. I didn't want to duplicate or confuse the error for this, which is a subset of the ways that could theoretically be encountered.
3. Valid. I'll address that and it'll probably end up removing the else anyway
4. Getting to where there's more and more duplication with the validation below for TGSM pointers. Maybe I can consolidate instead of duplicating

Reply:
I didn't end up merging any of the code since they were dealing with different variables at different stages anyway. I drill through GEP and also BC since we'd rather not rev the validator for an HLSL version that might have unions.